pkg/controllers/provisioning/scheduling/topologygroup.go

```diff
@@ -15,54 +15,283 @@ limitations under the License.
 package scheduling
 
 import (
+	"fmt"
 	"math"
 
+	"github.com/mitchellh/hashstructure/v2"
 	v1 "k8s.io/api/core/v1"
-	"k8s.io/apimachinery/pkg/util/sets"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/runtime"
+	utilsets "k8s.io/apimachinery/pkg/util/sets"
+
+	"github.com/aws/karpenter/pkg/apis/provisioning/v1alpha5"
+	"github.com/aws/karpenter/pkg/utils/sets"
+)
+
+type TopologyType byte
+
+const (
+	TopologyTypeSpread TopologyType = iota
+	TopologyTypePodAffinity
+	TopologyTypePodAntiAffinity
 )
 
-func NewTopologyGroup(pod *v1.Pod, constraint v1.TopologySpreadConstraint) *TopologyGroup {
-	return &TopologyGroup{
-		Constraint: constraint,
-		Pods:       []*v1.Pod{pod},
-		spread:     map[string]int{},
+func (t TopologyType) String() string {
+	switch t {
+	case TopologyTypeSpread:
+		return "topology spread"
+	case TopologyTypePodAffinity:
+		return "pod affinity"
+	case TopologyTypePodAntiAffinity:
+		return "pod anti-affinity"
 	}
+	return ""
 }
 
-// TopologyGroup is a set of pods that share a topology spread constraint
+// TopologyGroup is used to track pod counts that match a selector by the topology domain (e.g. SELECT COUNT(*) FROM pods GROUP BY(topology_ke
 type TopologyGroup struct {
-	Constraint v1.TopologySpreadConstraint
-	Pods       []*v1.Pod
-	// spread is an internal field used to track current spread
-	spread map[string]int // TODO(ellistarn) explore replacing with a minheap
+	// Hashed Fields
+	Key         string
+	Type        TopologyType
+	maxSkew     int32
+	namespaces  utilsets.String
+	rawSelector *metav1.LabelSelector // stored so we can easily hash
+	// Pod Index
+	owners map[types.UID]struct{} // Pods that have this topology as a scheduling rule
+	// Internal state
+	domains map[string]int32 // TODO(ellistarn) explore replacing with a minheap
+}
+
+func (t *TopologyGroup) Matches(namespace string, podLabels labels.Set) bool {
+	selector, _ := metav1.LabelSelectorAsSelector(t.rawSelector)
```

Comment:
thoughts on runtime.Must()?

I was even thinking about some sort of helper that was runtime.Must but for multiple return values
```
package hysteria

func Must2(value interface{}, err error) interface{} {
  runtime.Must(error)
  return value
}
```
```
return t.namespaces.Has(namespace) && hysteria.Must2(selector.Matches(podLabels)).(v1.Selector)
```

I wonder if this would be better with generics 🤔

Reply:
I'll just use runtime.Must on the error, found a few other places in topology listing code that ignored it and added it there too.